src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/PageDirectiveFeature.cs

```diff
@@ -3,20 +3,38 @@
 
 using System;
 using System.IO;
+using Microsoft.AspNetCore.Mvc.Razor.Extensions;
 using Microsoft.AspNetCore.Razor.Language;
 
 namespace Microsoft.AspNetCore.Mvc.RazorPages.Infrastructure
 {
     public static class PageDirectiveFeature
     {
+        private static readonly RazorEngine PageDirectiveEngine = RazorEngine.Create(builder =>
+        {
+            for (var i = builder.Phases.Count - 1; i >= 0; i--)
+            {
+                var phase = builder.Phases[i];
+                builder.Phases.RemoveAt(i);
+                if (phase is IRazorDocumentClassifierPhase)
+                {
+                    break;
+                }
+            }
+
+            RazorExtensions.Register(builder);
+            builder.Features.Add(new PageDirectiveParserOptionsFeature());
+        });
+
         public static bool TryGetPageDirective(RazorProjectItem projectItem, out string template)
         {
             if (projectItem == null)
             {
                 throw new ArgumentNullException(nameof(projectItem));
             }
 
-            return TryGetPageDirective(projectItem.Read, out template);
+            var sourceDocument = RazorSourceDocument.ReadFrom(projectItem);
+            return TryGetPageDirective(sourceDocument, out template);
         }
 
         public static bool TryGetPageDirective(Func<Stream> streamFactory, out string template)
```

Comment:
This should go away once we've have precompilation using the generated attribute.